Missed messages email cleanup

hackerkid opened the pull request:
Followup up of #11224
Fixes #11180

![screenshot from 2019-02-06 17-35-28](https://user-images.githubusercontent.com/7190633/52340406-b59b5500-2a35-11e9-83e8-41e93bf93080.png)

hackerkid commented:
@timabbott @rishig This is ready for a review.

rishig commented on templates/zerver/emails/missed_message.subject.txt:
```diff
@@ -1,8 +1,8 @@
+    {% elif stream_email_notify %} #{{ messages[0].header.plain }}
     {% endif %}
 {% else %}
     New missed message{{ message_count|pluralize }}
```
This should be changed to "New missed messages" (no message_count|pluralize)

rishig commented on templates/zerver/emails/missed_message.subject.txt:
```diff
@@ -1,8 +1,8 @@
     {% elif mention %} {{ sender_str }} mentioned you
-    {% elif private_message %} {{ sender_str }} sent you a message
-    {% elif stream_email_notify %} New messages in {{ messages[0].header.plain }}
+    {% elif private_message %} PMs with {{ sender_str }}
```
We should fix the commit message typo: Improvise -> Improve (on both the commits that have that).

rishig commented on templates/zerver/emails/missed_message.txt:
```diff
@@ -13,12 +13,16 @@ While you were away you received {{ message_count }} new{% if group_pm %} group
-Click here to log in to Zulip and view your new messages:
+{% if reply_to_zulip  %}
+Reply to this email directly, or view it in Zulip:
+
```
Let's remove these extra newlines (they'll show up in the sent message)

rishig commented on templates/zerver/emails/missed_message.source.html:
```diff
@@ -11,58 +11,32 @@
+        <a href="{{ realm_uri }}">Reply in Zulip</a>.<br>
+    {% endif %}
     {% if reply_warning %}
-    <p>Please do not reply to this automated message. To respond to the missed messages you have received, please log on to Zulip and send your replies there.</p>
```
We should remove this `reply_warning` section. (From here and the calling code).

rishig commented on templates/zerver/emails/missed_message.txt:
```diff
@@ -13,6 +9,15 @@ While you were away you received {{ message_count }} new{% if group_pm %} group
 
+—
+Manage email preferences:
+
```
These newlines show up in the final email, so should be suppressed (even though I see they were in the original?)

rishig commented:
Tim: small piece of context is that we'd like the names to be in-line, e.g.
`**Iago**: Lorem ipsum ..`

instead of 
```
_Iago_

Lorem ipsum ...
```

but it seemed too hard for the release deadline, so doing as followup.

timabbott commented:
OK; @hackerkid let me know when you've applied @rishig's changes.

hackerkid commented:
Updated :)

timabbott commented on templates/zerver/emails/missed_message.source.html:
```diff
@@ -45,9 +45,10 @@
-        <a href="{{ realm_uri }}">Click here to log in to Zulip and view your new messages.</a>
         {% if reply_to_zulip %}
-        Or just reply to this email.
+        Reply to this email directly, or <a href="{{ realm_uri }}">view it in Zulip</a>.
```
@rishig I don't think we should say "view it in Zulip" for something that isn't a direct link to the conversation; it's misleading.

rishig replied:
oh, I didn't realize the links were not links to the conversation itself. How hard would that be to change, now that we don't commingle different narrows?

timabbott replied:
We have nice functions for linking directly to the narrow (and in fact do so in the recipient headers in the previous design), so I don't think this should be a problem.

timabbott commented:
OK, posted a couple comments.  The one thing I'm unsure about after looking at your screenshot is the very long lines that this generates; do we want to set a maximum width style?  

In any case, this is probably ready to deploy on chat.zulip.org for testing.

rishig commented:
One thing that makes me feel better about no maximum width is that both Github emails and the gmail default don't have a maximum width.

timabbott commented:
@hackerkid can you rebase?  Now that 2.0 is out, I think we should plan to test-deploy this on chat.zulip.org this week.

hackerkid commented:
@timabbott What should be done about this?
https://github.com/zulip/zulip/pull/11465#pullrequestreview-201404192

hackerkid commented:
@timabbott @rishig This is ready for a review.

timabbott commented on zerver/lib/email_notifications.py:
```diff
@@ -163,34 +163,24 @@ def build_sender_payload(message: Message) -> Dict[str, Any]:
             header = "%s > %s" % (stream.name, message.topic_name())
             stream_link = stream_narrow_url(user_profile.realm, stream)
-            topic_link = topic_narrow_url(user_profile.realm, stream, message.topic_name())
+            narrow_link = topic_narrow_url(user_profile.realm, stream, message.topic_name())
```
This seems like a redundant call with what's above and should be deleted?  Might explain the surprising increase in number of database queries in this commit.

timabbott commented:
I merged the first commit.  A few comments on the rest:
* What happens to the styling of digest emails?  It looks from the code that this removes the message headers from those, which I'm not sure is intended (it seems like it might break things, since they still contain multiple recipients).
* Can you move the "Extract get_narrow_url" commit to the bottom of the branch? That seems mergable in any case (once the comment above is addressed).

timabbott commented:
(I'm test-deploying this on chat.zulip.org for feedback)

rishig commented:
cool! This is a big improvement. A few things I've noticed from the czo deploy
* Styling notes: We should remove our green link styling for these emails, and we should not set a footer font-size. 
* We should remove the period at the end of "Reply in Zulip" (and then for consistency, also from "Unsubscribe from missed message emails")
* We should remove the sender name for "X mentioned you" emails, since the sender name is in the title

And then a bigger thought (which could be followup)
* Github notifications use the person as the name of the sender, rather than "GitHub notifications" or similar. Now that I've experienced our strategy, I really appreciate how GitHub is doing it.
Our original issue with this strategy was that if you reply to the missed message email, we don't want this random email to be what gets entered in your address book (for the sender's name).
The way github handles this is they set the reply-to to something different from the sender. 
Is this an option for us?

rishig commented:
In situations where email replies don't work, let's actually make the second line of the footer:
This Zulip server is not configured to accept replies to missed messages ([help]). [Reply in Zulip]. 

(This also means we should keep the periods at the end of sentences in the footer.)

Tim, where should the help link point? Couldn't find anything obvious in the production docs. 

Other followups:
- [ ] We should say in the footer why you're getting the message
- [ ] We should change Mentions to look like stream or PM messages. 
- [ ] We should double check a reasonable thing happens if several people send messages in quick succession in a Group PM (e.g. that the missed message doesn't contain several messages from different people concatenated together).

timabbott commented:
For the link, https://zulip.readthedocs.io/en/latest/production/settings.html#miscellaneous-server-settings is probably the best link right now.  But if I recall correctly, @mateuszmandera and I have been talking about making dedicated page for the "Incoming email gateway" in ReadTheDocs; it seems like now would be a good time to do that migration.

mateuszmandera commented:
Yes, I think some good documentation for the email mirror would be useful - I have doing that on my list of TODOs.

hackerkid commented:
> What happens to the styling of digest emails? It looks from the code that this removes the message headers from those, which I'm not sure is intended (it seems like it might break things, since they still contain multiple recipients).

I have reverted this change.

> Can you move the "Extract get_narrow_url" commit to the bottom of the branch? That seems mergable in any case (once the comment above is addressed).

Updated.

> If it's not too hard to do the rebase, it'd be nice to move the Point "view in Zulip" link change to the beginning of the branch as well, since that one also seems like something we'll want even if we find the UI changes don't work out.

Merged this with "Improve the reply wording in missed_message" commit.

> Styling notes: We should remove our green link styling for these emails, and we should not set a footer font-size.

Updated.

> We should remove the sender name for "X mentioned you" emails, since the sender name is in the subject line

Sometimes the message body may have more than one sender. For example.

![Screenshot from 2019-06-18 23-50-49](https://user-images.githubusercontent.com/7190633/59709376-3e1bd480-9224-11e9-9bca-522b6cdff9bf.png)

Or do you mean something else?

> In situations where email replies don't work, let's actually make the second line of the footer:
Do not reply to this message. This Zulip server is not configured to accept replies to missed message emails ([help]). [Reply in Zulip].

Updated

This is ready for another round of of review :)

rishig commented:
Thanks @hackerkid! I'm collecting followups at
https://paper.dropbox.com/doc/Use-Zulip-from-your-email-roadmap--AfTwiBl0oM7dGAyQc38EAtxLAg-v6lim4PZSPZwmUGAai91I

since I think between all the email projects there's quite a bit going on. None of the followups on that list are necessary for getting this merged.

timabbott commented:
I merged the first commit with these changes to avoid doing a bunch of database queries twice:

```
diff --git a/zerver/lib/email_notifications.py b/zerver/lib/email_notifications.py
index e4573aa18e..e2e5bb85f6 100644
--- a/zerver/lib/email_notifications.py
+++ b/zerver/lib/email_notifications.py
@@ -1,5 +1,5 @@
 
-from typing import Any, Dict, Iterable, List, Tuple
+from typing import Any, Dict, Iterable, List, Optional, Tuple
 
 from confirmation.models import one_click_unsubscribe_link
 from django.conf import settings
@@ -163,19 +163,22 @@ def build_message_list(user_profile: UserProfile, messages: List[Message]) -> Li
                 'content': [build_message_payload(message)]}
 
     def message_header(user_profile: UserProfile, message: Message) -> Dict[str, Any]:
-        narrow_link = get_narrow_url(user_profile, message)
         if message.recipient.type == Recipient.PERSONAL:
+            narrow_link = get_narrow_url(user_profile, message)
             header = "You and %s" % (message.sender.full_name,)
             header_html = "<a style='color: #ffffff;' href='%s'>%s</a>" % (narrow_link, header)
         elif message.recipient.type == Recipient.HUDDLE:
-            disp_recipient = get_display_recipient(message.recipient)
-            assert not isinstance(disp_recipient, str)
-            other_recipients = [r['full_name'] for r in disp_recipient
+            display_recipient = get_display_recipient(message.recipient)
+            assert not isinstance(display_recipient, str)
+            narrow_link = get_narrow_url(user_profile, message,
+                                         display_recipient=display_recipient)
+            other_recipients = [r['full_name'] for r in display_recipient
                                 if r['id'] != user_profile.id]
             header = "You and %s" % (", ".join(other_recipients),)
             header_html = "<a style='color: #ffffff;' href='%s'>%s</a>" % (narrow_link, header)
         else:
             stream = Stream.objects.only('id', 'name').get(id=message.recipient.type_id)
+            narrow_link = get_narrow_url(user_profile, message, stream=stream)
             header = "%s > %s" % (stream.name, message.topic_name())
             stream_link = stream_narrow_url(user_profile.realm, stream)
             header_html = "<a href='%s'>%s</a> > <a href='%s'>%s</a>" % (
@@ -234,23 +237,31 @@ def build_message_list(user_profile: UserProfile, messages: List[Message]) -> Li
 
     return messages_to_render
 
-def get_narrow_url(user_profile: UserProfile, message: Message) -> str:
+def get_narrow_url(user_profile: UserProfile, message: Message,
+                   display_recipient: Optional[List[str]]=None,
+                   stream: Optional[Stream]=None) -> str:
     if message.recipient.type == Recipient.PERSONAL:
+        assert stream is None
+        assert display_recipient is None
         return personal_narrow_url(
             realm=user_profile.realm,
             sender=message.sender,
         )
     elif message.recipient.type == Recipient.HUDDLE:
-        disp_recipient = get_display_recipient(message.recipient)
-        assert not isinstance(disp_recipient, str)
-        other_user_ids = [r['id'] for r in disp_recipient
+        assert stream is None
+        if display_recipient is None:
+            display_recipient = get_display_recipient(message.recipient)
+        assert not isinstance(display_recipient, str)
+        other_user_ids = [r['id'] for r in display_recipient
                           if r['id'] != user_profile.id]
         return huddle_narrow_url(
             realm=user_profile.realm,
             other_user_ids=other_user_ids,
         )
     else:
-        stream = Stream.objects.only('id', 'name').get(id=message.recipient.type_id)
+        assert display_recipient is None
+        if stream is None:
+            stream = Stream.objects.only('id', 'name').get(id=message.recipient.type_id)
         return topic_narrow_url(user_profile.realm, stream, message.topic_name())
 
 def message_content_allowed_in_missedmessage_emails(user_profile: UserProfile) -> bool:
diff --git a/zerver/tests/test_digest.py b/zerver/tests/test_digest.py
index 4c9d8bbce0..897efd9f2c 100644
--- a/zerver/tests/test_digest.py
+++ b/zerver/tests/test_digest.py
@@ -47,7 +47,7 @@ class TestDigestEmailMessages(ZulipTestCase):
         with queries_captured() as queries:
             handle_digest_email(othello.id, cutoff)
 
-        self.assert_length(queries, 8)
+        self.assert_length(queries, 6)
 
         self.assertEqual(mock_send_future_email.call_count, 1)
         kwargs = mock_send_future_email.call_args[1]
@@ -106,7 +106,10 @@ class TestDigestEmailMessages(ZulipTestCase):
         with queries_captured() as queries:
             handle_digest_email(othello.id, cutoff)
 
-        self.assert_length(queries, 13)
+        # This can definitely be optimized; for both the huddle and
+        # stream cases, the get_narrow_url API ends up double-fetching
+        # some data because of how the functions are organized.
+        self.assert_length(queries, 9)
 
         self.assertEqual(mock_send_future_email.call_count, 1)
         kwargs = mock_send_future_email.call_args[1]
```

timabbott commented on zerver/lib/url_encoding.py:
```diff
@@ -27,7 +27,7 @@ def huddle_narrow_url(realm: Realm, other_user_ids: List[int]) -> str:
     return base_url + pm_slug
 
-def stream_narrow_url(realm: Realm, stream: Stream) -> str:
+def stream_narrow_url(realm: Realm, stream: Stream) -> str:  # nocoverage
```
This nocoverage isn't needed.

timabbott commented on zerver/tests/test_email_notifications.py:
```diff
@@ -186,7 +186,7 @@ def _test_cases(self, tokens: List[str], msg_id: int, body: str, email_subject:
             reply_to_addresses = [settings.EMAIL_GATEWAY_PATTERN % (u'mm' + t,) for t in tokens]
             reply_to_emails = [formataddr(("Zulip", address)) for address in reply_to_addresses]
-        else:
+        else:  # nocoverage: TODO
```
This nocoverage is in error; you shouldn't have removed the test for the case where we don't offer a direct reply feature.

I re-added the test (with a change to the text it checks for) and added this back.

timabbott commented:
I merged the rest of this, after the cleanups I mentioned above.
